### PR TITLE
mutation/mutation_compactor: validate the input stream

### PR DESCRIPTION
**Pull request by denesb**
The mutation compactor has a validator which it uses to validate the stream of mutation fragments that passes through it. This validator is supposed to validate the stream as it enters the compactor, as opposed to its compacted form (output). This was true for most fragment kinds except range tombstones, as purged range tombstones were not visible to
the validator for the most part.

This mistake was introduced by https://github.com/scylladb/scylladb/commit e2c9cdb576c2186719ce7181dd6619f1994e262e, which itself was a flawed attempt at fixing an error seen because purged tombstones were not terminated by the compactor.

This patch corrects this mistake by fixing the above problem properly: on page-cut, if the validator has an active tombstone, a closing tombstone is generated for it, to avoid the false-positive error. With this, range tombstones can be validated again as they come in.

The existing unit test checking the validation in the compactor is greatly expanded to check all (I hope) different validation scenarios.

### CODE REVIEW

**Comment by denesb** [comment]
This is a follow-up to https://github.com/scylladb/scylladb/pull/13749, requested by @avikivity.

**Comment by tchaikov** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1146/

looks like a genuine test failure from one of the tests added by this change.

**Comment by denesb** [comment]
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1146/
> 
> looks like a genuine test failure from one of the tests added by this change.

Yes, I'll have to look into it.

**Comment by avikivity** [comment]
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1146/
> 
> looks like a genuine test failure from one of the tests added by this change.

Unfortunately a rare event.

**Comment by denesb** [comment]
v2:
* When closing the active range tombstone, pass `tombstone{}` to the validator, instead of `{}`. The latter was evaluated to `std::nullopt` and didn't change the current tombstone, leading to test failure due to unclosed range tombstone at partition-end.

**Comment by denesb** [comment]
@scylladb/scylla-maint ping.

**Comment by denesb** [comment]
@scylladb/scylla-maint ping^2.

**Comment by denesb** [comment]
@scylladb/scylla-maint ping^3.

**Comment by denesb** [comment]
Rebased.

**Comment by nyh** [comment]
> @scylladb/scylla-maint ping^3.

Sorry about that, I'll start reviewing it this PR now. 
But as a general request, let's start getting more developers involved in reviews, not just the maintainers. If there's somebody with whom you previously discussed this idea (e.g., @avikivity who you said asked for this patch) or who previously wrote code in this area, they should review this PR. Only then should the general list of maintainers be called in.

**Review by nyh — APPROVED** [review]

**Comment by mykaul** [comment]
> > @scylladb/scylla-maint ping^3.
> 
> Sorry about that, I'll start reviewing it this PR now. But as a general request, let's start getting more developers involved in reviews, not just the maintainers. If there's somebody with whom you previously discussed this idea (e.g., @avikivity who you said asked for this patch) or who previously wrote code in this area, they should review this PR. Only then should the general list of maintainers be called in.

We could look into implementing some of of the features available @ https://docs.github.com/en/organizations/organizing-members-into-teams/managing-code-review-settings-for-your-team

**Comment by nyh** [comment]
@mykaul personally I would prefer to see people want to review, not be assigned random PRs to review by software robots. I honestly think that people should usually be working on something which interests another person (perhaps the other person asked for this fix (as was the case here), or assigned you the fix, or wrote the code you are fixing, or working with you on the bigger feature that you are now improving. Then that other person can be asked directly to review the patch before inviting some team or the maintainers to review. I guess that if no other such person exists, calling in a team might be helpful.

**Comment by mykaul** [comment]
> @mykaul personally I would prefer to see people want to review, not be assigned random PRs to review by software robots. I honestly think that people should usually be working on something which interests another person (perhaps the other person asked for this fix (as was the case here), or assigned you the fix, or wrote the code you are fixing, or working with you on the bigger feature that you are now improving. Then that other person can be asked directly to review the patch before inviting some team or the maintainers to review. I guess that if no other such person exists, calling in a team might be helpful.

Agreed, but I'd like a bot to be called in for rescue if after X days (2?) no one volunteered to review.

**Pull request merged**
